src/SimplePromiseQueue.ts

```diff
@@ -1,25 +1,30 @@
 // tslint:disable: variable-name
 export default class SimplePromiseQueue {
   private readonly _queue: Array<() => Promise<void>> = []
-  private _flushing = false
+  private _runner: Promise<void> | null = null
 
   public enqueue(task: () => Promise<void>) {
     this._queue.push(task)
-    if (!this._flushing) { return this.flushQueue() }
-    return Promise.resolve()
+    this.join()
   }
 
-  private flushQueue() {
-    this._flushing = true
+  public async join(): Promise<void> {
+    if (this._runner) {
+      return await this._runner
+    }
+    this._runner = this.joinImpl()
+    await this._runner
+    this._runner = null
+  }
 
-    const chain = (): Promise<void> | void => {
+  private async joinImpl(): Promise<void> {
```

Comment:
what is Impl?

Reply:
For implementation. I can rename this to be more clear.